android/src/main/java/com/reactnativestripesdk/StripeSdkCardView.kt

```diff
@@ -193,14 +193,21 @@ class StripeSdkCardView(private val context: ThemedReactContext) : FrameLayout(c
       cardDetails["brand"] = null
       cardDetails["last4"] = null
     }
+    sendCardDetailsEvent()
+  }
+
+  private fun sendCardDetailsEvent() {
     mEventDispatcher?.dispatchEvent(
       CardChangedEvent(id, cardDetails, mCardWidget.postalCodeEnabled, cardParams != null, dangerouslyGetFullCardDetails))
   }
 
   private fun setListeners() {
     mCardWidget.setCardValidCallback { isValid, _ ->
       if (isValid) {
-        onCardChanged()
+        onValidCardChange()
+      } else {
+        cardParams = null
+        cardAddress = null
```

Comment:
Hi @michelleb-stripe I'm taking this over from Arek but I'm not sure I get this properly. Can you please help me understanding the problem?

Reply:
@souhe Let me try to explain more.  The setExpiryDateTextWatcher and setPostalCodeTextWatchedr called onCardChanged, which will get the paymentMethodCreateParams without first checking to see if they are valid.   Because of how a query on the paymentMethodCreateParams works it will change the focus on the widget, to the first field that is invalid.

paymentMethodCreateParams should only be called from the CardValidCallback.

Reply:
@michelleb-stripe I'm afraid I'm still confused :D Are you describing your previous comment or this whole issues?

If the first one then I'm not getting why calling `sendCardDetailsEvent` from `setExpiryDateTextWatcher` or `setPostalCodeTextWatcher` would call `paymentMethodCreateParams`. 

If the second one I'm still not sure what this means - `I think that would mean that the CardValidCallback would always need to be followed by a TextWatcher. I am not sure this is a guarantee we can make`. Why is that true? If I understand the code created by Arek properly we are sending `sendCardDetailsEvent` on every change from each watcher so that is why there is no need to send this even once again from cardValidCallback (when card is invalid)

Reply:
Sorry for the delay.   Looking at it again it looks good to me.